llvm/lib/Target/WebAssembly/WebAssemblyFrameLowering.cpp

```diff
@@ -147,7 +147,14 @@ bool WebAssemblyFrameLowering::needsPrologForEH(
 /// Unlike a machine stack pointer, the wasm user stack pointer is a global
 /// variable, so it is loaded into a register in the prolog.
 bool WebAssemblyFrameLowering::needsSP(const MachineFunction &MF) const {
-  return needsSPForLocalFrame(MF) || needsPrologForEH(MF);
+  auto &MRI = MF.getRegInfo();
+  // llvm.stacksave can explicitly read SP register and it can appear without
+  // dynamic alloca.
+  bool hasExplicitSPUse =
```

Comment:
Maybe this should instead go in `needsSPForLocalFrame()`? That function has other logic related to whether an SP is needed based on the current function.

Reply:
Actually, even more than just where related code is located, `needsSPForLocalFrame()` is also used by `needsSPWriteback()`. I guess if the stacksave and stackrestore are properly paired, then in theory an SP writeback at the end of the function wouldn't be needed. But maybe it makes sense to have it anyway, just to be safe.

Reply:
That sounds reasonable to me :+1: Updated.